Express y grid > 2^16 in terms of z grid

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121554

CUDA has a max y_grid of 65535. If we're computing larger than that we can compose it in terms of z grid, which is currently unused in inductor codegen.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Express y grid > 2^16 in terms of z grid

[ghstack-poisoned]

eellison committed:
Update on "Express y grid > 2^16 in terms of z grid"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Express y grid > 2^16 in terms of z grid"


CUDA has a max y_grid of 65536. If we're computing larger than that we can compose it in terms of z grid, which is currently unused in inductor codegen.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Express y grid > 2^16 in terms of z grid"


CUDA has a max y_grid of 65536. If we're computing larger than that we can compose it in terms of z grid, which is currently unused in inductor codegen.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Express y grid > 2^16 in terms of z grid"


CUDA has a max y_grid of 65536. If we're computing larger than that we can compose it in terms of z grid, which is currently unused in inductor codegen.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1096,6 +1096,13 @@ def scalar_code(self, value):
+        if self.grid_dim == 1 and config.triton.max_tiles <= 2 and not (
+            isinstance(self.numel, int) and self.numel <= MAX_Y_GRID
+        ):
+            key = f'{key} * {f"tl.program_id({self.grid_dim + 1})"}'
```
Maybe use one f-string?

aakhundov commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -1436,11 +1436,28 @@ def get_grid_dim(numel, block):
+            z_grid = get_grid_dim(znumel, meta.get("ZBLOCK", None))
+            torch._check(
+                y_grid <= 65535,
+                lambda: f"Generated y grid behind 2^16 ({y_grid}) not supported with z dimension present. File issue",
```
Maybe s/behind/larger than/ for clarity? Although likely a non-native-speaker issue on my side :)

jansel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -1436,11 +1436,28 @@ def get_grid_dim(numel, block):
+        x_grid = get_grid_dim(xnumel, meta.get("XBLOCK", 1))
+        y_grid = get_grid_dim(ynumel, meta.get("YBLOCK", None))
+
+        MAX_Y_GRID = 65536
```
move to global scope (and remove the other copy)

jansel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -1436,11 +1436,28 @@ def get_grid_dim(numel, block):
+        else:
+            z_grid = get_grid_dim(znumel, meta.get("ZBLOCK", None))
+            torch._check(
+                y_grid <= 65535,
```
```suggestion
                y_grid < MAX_Y_GRID,
```

jansel reviewed (dismissed):
Tests are failing

aakhundov commented:
> Tests are failing

Sorry, was too hasty to accept.

eellison committed:
Update on "Express y grid > 2^16 in terms of z grid"


CUDA has a max y_grid of 65536. If we're computing larger than that we can compose it in terms of z grid, which is currently unused in inductor codegen.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Express y grid > 2^16 in terms of z grid"


CUDA has a max y_grid of 65535. If we're computing larger than that we can compose it in terms of z grid, which is currently unused in inductor codegen.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge -f "infra failure"
